### PR TITLE
Changelog improvements for 3.3

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Improve some changelog entries:

* Explain "Codegen 1.1", and add a "Requirements change" mentioning `jsonschema` and `basic.requirements.txt` (we had several support requests about that).
* Clarify various entries, in particular around TLS 1.3.
* Remove discussions of internal changes. Changelog entries are for users, not for maintainers.
* Various wording improvements, wrap long lines, etc.

## Gatekeeper checklist

- [x] **changelog** N/A
- [ ] **backport** TODO
- [x] **tests** N/A

### CODE REVIEW

File: ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
 Features
-   * When GnuTLS/Openssl server is configured in TLS 1.2 mode with a certificate
-     declaring an RSA public key and Mbed TLS is configured in hybrid mode, if
-     `rsa_pss_rsae_*` algorithms are before `rsa_pkcs1_*` ones in this list then
-     the GnuTLS/Openssl server chooses an `rsa_pss_rsae_*` signature algorithm
-     for its signature in the key exchange message. As Mbed TLS 1.2 does not
-     support them, the handshake fails. Add `rsa_pss_rsae_*` support for TLS 1.2
-     to resolve the compitablity issue.
+   * Support rsa_pss_rsae_* signature algorithms in TLS 1.2.
+Bugfix
```
Comment (yanesca) [inline ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt:3]:
```suggestion
   * Support rsa_pss_rsae_* signature algorithms in TLS 1.2.
     Bugfix
```

**Comment by ronald-cron-arm** [comment]
I've looked at the changes related to TLS 1.3, namely the changes in add-rsa-pss-rsae-support-for-tls12.txt, fix_tls13_session_resumption_fail_when_hostname_is_not_localhost.txt and tls13-misc.txt. This looks good to me.

File: ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt (hunk 1 above)
Comment (gilles-peskine-arm) [reply ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt:3]: I don't understand this suggested change. “Bugfix” is the title of a section, it needs to be at the start of the line.

Comment (yanesca) [reply ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt:3]: My bad, sorry
```suggestion
   * Support rsa_pss_rsae_* signature algorithms in TLS 1.2.
   
Bugfix
```
Although I am not sure if this makes any difference to the changelog assembly script.

File: ChangeLog.d/fix-tls12server-sent-sigalgs.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 Bugfix
-    * Fix a bug whereby the the list of signature algorithms sent as part of the
-      TLS 1.2 server certificate request would get corrupted, meaning the first
-      algorithm would not get sent and an entry consisting of two random bytes
-      would be sent instead. Found by Serban Bejan and Dudek Sebastian.
+    * Fix a bug whereby the the list of signature algorithms sent as part of
```
Comment (tom-daubney-arm) [inline ChangeLog.d/fix-tls12server-sent-sigalgs.txt:2]: "Fix a bug whereby the **the** list" - Remove the second "the".

File: ChangeLog.d/fix_cmake_gen_files.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
 Bugfix
    * Fix an issue with CMake builds in releases with GEN_FILES turned off,
-     whereby missing generated files could be turned into symlinks to themselves.
+     whereby missing generated files could be turned into symlinks to
```
Comment (tom-daubney-arm) [inline ChangeLog.d/fix_cmake_gen_files.txt:3]: Similar to before. I prefer "omitted" to "missing" here. Please consider changing it.

**Review by tom-daubney-arm — CHANGES_REQUESTED** [review]
A few small gramatical nits to iron out but mostly looking good.

File: ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt (hunk 1 above)
Comment (gilles-peskine-arm) [reply ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt:3]: The script doesn't care about blank lines. Any line that starts with a letter is a section header.

File: ChangeLog.d/fix_cmake_gen_files.txt [hunk 4]
```diff
@@ -1,3 +1,3 @@
 Bugfix
-   * Fix an issue in releases with GEN_FILES turned off whereby missing
-     generated files could be turned into symlinks to themselves.
+   * Fix an issue with CMake builds in releases with GEN_FILES turned off,
+     whereby missing generated files could be turned into symlinks to themselves.
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: I thought I understand what this means, and I don't understand what “omitted generated files” would be. I'll check what the bug was exactly and reformulate.

File: ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt [hunk 5]
```diff
@@ -1,8 +1,8 @@
...
+   * Fix an interoperability failure between an Mbed TLS client with both
+     TLS 1.2 and TLS 1.3 support, and a TLS 1.2 server such as GnuTLS or
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/add-rsa-pss-rsae-support-for-tls12.txt:5]: I don't think it really matters which software implements the server here, so I'll remove this information.

File: ChangeLog.d/fix_cmake_gen_files.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: The issue is https://github.com/Mbed-TLS/mbedtls/pull/6086. It's about is a generated file, which is shipped in the build, is missing from the working tree. It's not about something being “omitted”. I'll reformulate, but I can't think of a better word than “missing”.

File: ChangeLog.d/fix_cmake_gen_files.txt (hunk 3 above)
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt:3]: That's the same as https://github.com/Mbed-TLS/mbedtls/pull/6689/files/d9d992f6715731cff41293611437c5c823a19c90..16ebf97cbb2798c27233807cb36adff7af100644#diff-580aa7c89d3f4c4c63d50bcd197ec7ae9301632754bc8f92d22fdae5c3f36414

File: ChangeLog.d/fix_cmake_gen_files.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (tom-daubney-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: I see what you mean. Perhaps "lacking" instead of "missing"?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: You mean “if the working directory was lacking a generated file, this file could be turned into a symlink of itself”? I find this harder to understand.

[the remark this answers is not in the document]
Comment (tom-daubney-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: Yeah that is what I mean. Maybe to tweak even further: “if the working directory was lacking a generated file, **said** file could be turned into a symlink of itself”

I am happy for you to go with your gut here. I personally find the above easiest to read, but I am also totally aware that this is subjective. Go with your gut and I won't hold the PR up over this. Doing my re-review now.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt]: I'm not planning any further changes until I hear from the second reviewer.

**Review by tom-daubney-arm — DISMISSED** [review]
Thanks for the changes. Everything looking good to me now.

**Comment by gilles-peskine-arm** [comment]
I've rewritten the history to not touch the PKCS#7 entry, since it's being removed in https://github.com/Mbed-TLS/mbedtls/pull/6709.

**Review by yanesca — APPROVED** [review]
LGTM

File: ChangeLog.d/tls13-misc.txt [hunk 6]
```diff
@@ -1,9 +1,8 @@
...
-   * Mbed TLS supports TLS 1.3 key establishment via pre-shared keys,
-     pre-shared keys provisioned externally or via the ticket mechanism
-     (session resumption).
-     The MBEDTLS_SSL_SESSION_TICKETS configuration option controls the support
-     for the ticket mechanism.
-     MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_xxx_ENABLED configuration options
-     have been introduced to control the support for the three possible
-     TLS 1.3 key exchange modes.
+   * Mbed TLS now supports TLS 1.3 key establishment via pre-shared keys.
+     The pre-shared keys can provisioned externally or via the ticket
```
Comment (tom-daubney-arm) [inline ChangeLog.d/tls13-misc.txt:3]: Missing word, "be" --> "The pre-shared keys can **be** provisioned externally..."

**Review by tom-daubney-arm — CHANGES_REQUESTED** [review]
Found one remaining typo. Once fixed this will get an approval from me.

File: ChangeLog.d/fix_hard_link_across_drives.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Bugfix
+   * Fix a build issue on Windows using CMake where the source and build
+     directory could not be on different drives. Fixes #5751.
```
Comment (tom-cosgrove-arm) [inline ChangeLog.d/fix_hard_link_across_drives.txt:3]: err, should be `source and build directories` (plural)

File: ChangeLog.d/fix_cmake_gen_files.txt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Fix an issue with in-tree CMake builds in releases with GEN_FILES
+     turned off: if a shipped file was missing from the working directory,
+     it could be turned into a symbolic link to itself.
```
Comment (tom-cosgrove-arm) [inline ChangeLog.d/fix_cmake_gen_files.txt:4]: I still don't understand this.

Is it "if a generated file wasn't present in the working directory, a symbolic link pointing to itself would be created"?

Comment (tom-daubney-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt:4]: That's how I have parsed it, yes.

Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt:4]: That is my understanding of https://github.com/Mbed-TLS/mbedtls/pull/6086 indeed. Possibly only under some circumstances (I haven't investigated), hence “could”.

**Comment by gilles-peskine-arm** [comment]
I fixed two more reported issues (none applicable to 2.28).

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: ChangeLog.d/fix_cmake_gen_files.txt (hunk 8 above)
Comment (daverodgman) [reply ChangeLog.d/fix_cmake_gen_files.txt:4]: I think Tom's point is that if it's shipped, it's not missing? You could say "if a file that should have been shipped was missing..." but it's getting a bit wordy? IMO the current wording isn't optimal but it's sufficient.

Comment (gilles-peskine-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt:4]: What can happen in practice is that a file was shipped, but is not present at this time (because the user removed it, e.g. by running `make neat`). The file would not be “omitted” for anywhere, but it would be “missing” from the build tree.

Comment (tom-cosgrove-arm) [reply ChangeLog.d/fix_cmake_gen_files.txt:4]: that's why I used "not present", which covers all of those things, and means just that's it's not there
